bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/url/URLStreamHandlerFactoryImpl.java

```diff
@@ -54,10 +53,7 @@ public class URLStreamHandlerFactoryImpl extends MultiplexingFactory implements
 	 */
 	public URLStreamHandlerFactoryImpl(BundleContext context, EquinoxContainer container) {
 		super(context, container);
-
-		proxies = new Hashtable<>(15);
-		handlerTracker = new ServiceTracker<>(context, URLSTREAMHANDLERCLASS, null);
```

Comment:
> I assume if someone is doing operations with a URL for a protocol then we will need the real service right away. 

The problem is that the "real" impl can come and go and we need to take account for that so the change seems a lot easier than working with different trackers and raw service references.

Reply:
> The problem is that the "real" impl can come and go
I understand that.  I was pointing out that to fix the issue #289 all you need to do is make the first tracker not be eager in getting the service or get rid of it altogether.  I don't think making the "inner" trackers for each protocol be lazy also has that many advantages unless you have many services that implement the same protocol.

> and we need to take account for that so the change seems a lot easier than working with different trackers and raw service references.

Doesn't seem that much easier, but since you already did the work and I don't think it makes it conceptually more complicated then I am fine with it.  It does get rid of the one "main" tracker at least.  But I think we could have done that without introducing the `LazyURLStreamHandlerService`